### PR TITLE
Fix performance regressions in the P2P network

### LINKED ISSUE #2272: P2P: node not pruning enough neighbors
On Xenon, I observed 182 outbound neighbors:

```bash
➜ curl -s http://127.0.0.1:20443/v2/neighbors | jq '.outbound[].public_key_hash' | sort | uniq | wc -l
     182
```

Per @jcnelson, [this seems too high](https://discord.com/channels/621759717756370964/730814571517968436/796107541703688222):

```
The node should have been pruning its inbound and outbound neighbor sets
Will look more into that.  182 is great to see, but it's really high
```

### PR DESCRIPTION
**Pull request by jcnelson**
This fixes #2272, #2281, #2280, #2022, and #2320.  It makes it so we prune connections _quickly_, we set sensible neighbor counts, we don't spend as much time fetching redundant block inventory data, and we limit the number of writes we do on the frontier DB per P2P state machine pass.  I have confirmed multiple times that I can boot up a Xenon node with this PR.

The PR sets and enforces a maximum number of outbound connections to 16 peers, but it permits up to 750 inbound p2p connections (and decouples this count from the number of inbound HTTP requests).  To ensure that a sparse peer set doesn't allow the node to get cut off from nodes that are likely to have anchor blocks (i.e. miners), the PR makes it so any bootstrap nodes the node has configured will _always_ be neighbors.  Moreover, the PoX sync watchdog will not permit the node to fetch the next reward cycle until it has confirmed that at least one bootstrap node has been scanned.  In addition, this PR makes it so that the neighbor-walk logic always tries to keep a bootstrap node as a neighbor, and makes it so the inventory synchronization logic always assumes that a bootstrap node that NACKs a request for inventory data due to an invalid consensus hash is correct  in doing so -- i.e. if this happens, then the local node, not the remote node, is assumed to have offered a bad consensus hash in the request (instead of the remote node having incomplete chainstate).

This PR effectively makes it so that a node's bootstrap nodes are always trusted to have exactly the right anchor blocks.  When we deploy this PR, we should be sure to have enough seed nodes that each miner gets reliably connected as an inbound peer to at least one seed node (so the network will stay in-sync and nodes can bootstrap without deep PoX reorgs).  Once we have #1805 ready, then this trust assumption will no longer be necessary -- #1805 will make it so nodes will be able to use the burnchain state to infer which neighbors' inventory NACKs are appropriate.

Please test with a Xenon node and verify that you can reach the chain tip without deep PoX reorgs (i.e. no more than 1 or 2 reward cycles).  In debug mode, should see the `PoxId(..)` occurrences in the log file grow with `true`'s at a linear pace.  At the time of this writing, there should be no `false`s.

### CODE REVIEW

**Comment by jcnelson** [comment]
This is now ready for review.

File: testnet/stacks-node/src/syncctl.rs [hunk 1]
```diff
@@ -395,7 +395,7 @@ impl PoxSyncWatchdog {
             // burnchain has advanced since the last call
             self.last_burnchain_height = burnchain_height;
             debug!("PoX watchdog: Wait for at least one inventory state-machine pass...");
-            self.relayer_comms.wait_for_inv_sync_pass(self.max_samples);
+            self.relayer_comms.wait_for_inv_sync_pass(3600); // wait an hour at most
```
Comment (kantai) [inline testnet/stacks-node/src/syncctl.rs:398]: Isn't waiting an hour way too long? Or is the goal here that if a node cannot complete a full inventory sync pass, it shouldn't process burnchain blocks at all?

**Review by kantai — APPROVED** [review]
This LGTM!

**Comment by kantai** [comment]
These changes appear to have broken the `neon_integrations` tests -- it may be related to the `pox_sync_watchdog` changes. Can you fix this before merging?

**Comment by jcnelson** [comment]
Yes, I'll fix them before merging.

File: testnet/stacks-node/src/syncctl.rs (hunk 1 above)
Comment (jcnelson) [reply testnet/stacks-node/src/syncctl.rs:398]: The latter.

Comment (lgalabru) [reply testnet/stacks-node/src/syncctl.rs:398]: Should this value be coming from config?

File: src/net/p2p.rs [hunk 2]
```diff
@@ -3004,14 +3030,16 @@ impl PeerNetwork {
...
+                    do_prune = true;
+                    /*
+                    do_prune = match self.walk {
+                        Some(ref mut walk) => {
+                            walk.do_prune = false;
+                            true
+                        }
+                        None => true
+                    };
+                    */
```
Comment (lgalabru) [inline src/net/p2p.rs:3042]: Should we add a TODO or remove that snippet?

**Review by lgalabru — APPROVED** [review]
Looks good to me, thank you @jcnelson!

File: src/net/p2p.rs (hunk 2 above)
Comment (jcnelson) [reply src/net/p2p.rs:3042]: The snippet should be removed; will do.

**Comment by jcnelson** [comment]
Just waiting for the `nettest` to pass.  I'm also running it locally because I want to be reasonably sure this doesn't break anything.

**Pull request merged**